src/Backend/Haskell.idr

```diff
@@ -1,68 +1,107 @@
 module Backend.Haskell
 
-import Data.Vect
 import Control.Monad.State
+import Text.PrettyPrint.WL
 
 import Backend.Utils
 import Types
 import Typedefs
 
-%default partial
+import Data.Vect
+
+%default total
 %access public export
 
-guardPar : String -> String
-guardPar str = if any isSpace $ unpack str then parens str else str
+data HType : Type where
+  H0     :                               HType
+  H1     :                               HType
+  HProd  :         Vect (2 + k) HType -> HType
+  HVar   :         Name               -> HType
+  HParam : Name -> Vect k HType       -> HType
+
+data Haskell : Type where
+  Synonym : Decl -> HType                -> Haskell
+  ADT     : Decl -> Vect k (Name, HType) -> Haskell
 
-nameWithParams : Name -> Env n -> String
-nameWithParams name e = withSep " " id (uppercase name::map lowercase (getFreeVars e))
+-- Render a name applied to a list of arguments exactly as written (arguments need to be previously parenthesized, if applicable).
```

Comment:
I think many of the comments could (and probably should) be turned into doc comments.